src/psyclone/tests/dynamo0p3_test.py

```diff
@@ -1679,68 +1687,98 @@ def test_dyninvoke_first_access():
+    assert args_out == []
+    args_in = [arg.declaration_name for arg in args['in']]
+    assert args_in == ['a']
 
 
 def test_dyninvoke_uniq_declns_intent_int():
-    ''' tests that DynInvoke.unique_declns_by_intent() returns the correct
-    list of arguments for gh_integer '''
+    ''' Tests that DynInvoke.unique_declns_by_intent() returns the correct
+    list of arguments for 'gh_integer' argument type. '''
     _, invoke_info = parse(os.path.join(BASE_PATH,
                                         "1.7_single_invoke_2scalar.f90"),
                            api=TEST_API)
     psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
     args = psy.invokes.invoke_list[0].unique_declns_by_intent("gh_integer")
-    assert args['inout'] == []
-    assert args['out'] == []
-    assert args['in'] == ['istep']
+    args_inout = [arg.declaration_name for arg in args['inout']]
+    assert args_inout == []
+    args_out = [arg.declaration_name for arg in args['out']]
+    assert args_out == []
+    args_in = [arg.declaration_name for arg in args['in']]
+    assert args_in == ['istep']
 
 
 def test_dyninvoke_uniq_declns_intent_ops():
-    ''' tests that DynInvoke.unique_declns_by_intent() returns the correct
-    list of arguments for operator arguments '''
+    ''' Tests that DynInvoke.unique_declns_by_intent() returns the correct
+    list of arguments for operator arguments. '''
     _, invoke_info = parse(os.path.join(BASE_PATH,
                                         "4.4_multikernel_invokes.f90"),
                            api=TEST_API)
     psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
     args = psy.invokes.invoke_list[0].unique_declns_by_intent("gh_operator")
-    assert args['inout'] == []
-    assert args['out'] == ['op']
-    assert args['in'] == []
+    args_inout = [arg.declaration_name for arg in args['inout']]
+    assert args_inout == []
+    args_out = [arg.declaration_name for arg in args['out']]
+    assert args_out == ['op']
+    args_in = [arg.declaration_name for arg in args['in']]
+    assert args_in == []
+
+
+def test_dyninvoke_uniq_declns_intent_cma_ops():
+    ''' Tests that DynInvoke.unique_declns_by_intent() returns the correct
+    list of arguments for columnwise operator arguments. '''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "20.5_multi_cma_invoke.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    args = psy.invokes.invoke_list[0]\
+        .unique_declns_by_intent("gh_columnwise_operator")
+    args_inout = [arg.declaration_name for arg in args['inout']]
+    assert args_inout == ['cma_opc']
+    args_out = [arg.declaration_name for arg in args['out']]
+    assert args_out == ['cma_op1']
+    args_in = [arg.declaration_name for arg in args['in']]
+    assert args_in == ['cma_op1', 'cma_opb']
```

Comment:
Unless I'm missing something, we should not have `cma_op1` listed as having both intent 'out' and intent 'in'?

Reply:
In the test example used here, `20.5_multi_cma_invoke.f90`, `cma_op1` is "read" arg in `columnwise_op_app_kernel_type` call and "write" arg in `columnwise_op_mul_kernel_type` call. This was about the only example I found with multiple CMA operator calls. I added comments in the test to explain this.

Reply:
Yes, you're right and I think the problem is deeper than that. We have 
```
        # sum_args behave as if they are write_args from
        # the PSy-layer's perspective.
        write_args += sum_args
        # readwrite_args behave in the same way as inc_args
        # from the perspective of first access and intents
        inc_args += readwrite_args
```
so byt his logic checks for "write" args should reject anything else than `AccessType.WRITE, AccessType.SUM`, whereas checks for "inc" should reject anything else than `AccessType.INC, AccessType.READWRITE`. However, now we have
```
        for arg in inc_args:
            ...
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access != AccessType.WRITE:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
            ...
```
and
```
        for arg in write_args:
            ...
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access == AccessType.READ:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
            ...
```
No wonder Coverage shows red for these lines. I am happy to create an issue for this.

Reply:
Yes, you're right and I think we need to revisit the logic there. We have
```
        # sum_args behave as if they are write_args from
        # the PSy-layer's perspective.
        write_args += sum_args
        # readwrite_args behave in the same way as inc_args
        # from the perspective of first access and intents
        inc_args += readwrite_args
```
There are first checks for "inc" and "readwrite" args 
```
        for arg in inc_args:
            ...
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access != AccessType.WRITE:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
            ...
```
and then for pure writers
```
        for arg in write_args:
            # For every argument that is written to by at least one kernel,
            # identify the type of the first access - if it is read
            # or inc'd before it is written then it must have intent(inout).
            # However, we deal with inc and readwrite args separately so we
            # do not consider those here.
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access == AccessType.READ:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
```
However, I think "However, we deal with inc and readwrite args separately so we do not consider those here" is not fool-proof as there is a possibility that a "write" arg can have a  "read"-like first access access other than pure "READ"'. I tested this hypothesis with
```
        for arg in write_args:
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access in (AccessType.READ, AccessType.READWRITE, AccessType.INC):
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
```
Now 'cma_op1' is correctly placed in `inout` and `in` lists. 

I created an issue and referred it is as TODO in `psyGen` and this test.

Reply:
Yes please.

Reply:
Created the issue and referred it accordingly.